rfd/00xx-snowflake-support.md

```diff
@@ -0,0 +1,171 @@
+
+```yaml
+db_service:
+  enabled: "yes"
+  databases:
+    - name: "snowflake"
+      description: "❄ Snowflake database"
+      protocol: "snowflake"
+      uri: "im12345.us-east-2.aws.snowflakecomputing.com"
+```
+
+The next step would be to add teleport CA to the Snowflake user account. `tctl` could learn to export the required keys 
+in the same way as it does today for different databases (`tctl auth sign` command). The exported key must then be added 
+to Snowflake by an administrator. The procedure of adding the key to Snowflake is described [here](https://docs.snowflake.com/en/user-guide/key-pair-auth.html#step-4-assign-the-public-key-to-a-snowflake-user).
+
+### Database access support
+
+`tsh db login` should export the user private key currently stored for example in 
+`~/.tsh/keys/proxy.example.com/alice` in PCKS8 format (that's the only format supported by `snowsql` and some SDKs).
+That key should be used to sign a "Teleport JWT". Client then should send the request with generated JWT to authenticate 
+with Snowflake. Teleport can intercept HTTP communication between a client and the Snowflake. The first request send by 
+the client should be to the `/session/v1/login-request` endpoint which contains all login information as username, JWT. 
+This data can be used to authenticate user in Teleport. JWT token prepared by the client can be then read and replaces by Snowflake JWT
+generated by Teleport DB service. Teleport should use Database CA to sign JWT at that point to sign the request instead 
+of User CA to be able to authenticate with the key generated by `tctl auth sign` and imported to Snowflake. 
+In this way the real token is not accessible by a user and Teleport can still validate if the request 
+comes from a trusted source.
+
+```mermaid
+sequenceDiagram
+    participant client
+    participant Teleport
+    participant Snowflake
+    client->>Teleport: HTTP /session/v1/login-request
+    Teleport->>Teleport: RBAC check, genererate and replace JWT
+    Teleport->>Snowflake: HTTP /session/v1/login-request
+    Snowflake->>Teleport: Authentication Token
+    Teleport->>Teleport: Generate Teleport Auth token
+    Teleport->>client: Send Teleport Auth token
+```
+
+On every request Teleport can replace the "Teleport Auth token" with Snowflake Auth token:
+
+```mermaid
+sequenceDiagram
+    participant client
+    participant Teleport
+    participant Snowflake
+    client->>Teleport: HTTP /queries/v1/query-request
+    Teleport->>Teleport: Replace Auth header
+    Teleport->>Snowflake: HTTP /queries/v1/query-request
+    Snowflake->>Teleport: Response
+    Teleport->>client: Forward the response
+```
+
+With this approach the real session token never leaves the Teleport and any client is not able to bypass Teleport
+by extracting the session token.
+Snowflake session token needs to be refreshed every X minutes. For that purpose Teleport would have to internally refresh
```

Comment:
Just leaving a remainder comment that we should ensure that Snowflake API doesn't have any Quota for token refreshing operation.

Reply:
I hope that maybe Snowflake will share some docs with us, so we know what the limit is. Until then I can just add, that there maybe some limit 😐